### PR TITLE
support for WS-Trust 2005

### PR DESCRIPTION
**Pull request by kpanwar**
support for WS-Trust 2005
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/223%23discussion_r35997221%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/223%23discussion_r35997813%22%5D%2C%20%22comments%22%3A%20%7B%22Pull%205021a1d4968aa64a1f235a2124fb94efd3b8bc53%20src/ADAL.Common/MexParser.cs%2052%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/223%23discussion_r35997813%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Why%20is%20this%20randomized%3F%22%2C%20%22created_at%22%3A%20%222015-07-31T17%3A51%3A50Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7756476%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aj-michael%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20src/ADAL.Common/MexParser.cs%3AL85-110%22%7D%2C%20%22Pull%205021a1d4968aa64a1f235a2124fb94efd3b8bc53%20src/ADAL.Common/WsTrustRequest.cs%20163%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/223%23discussion_r35997221%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22typo%20for%20expiryTimeString%22%2C%20%22created_at%22%3A%20%222015-07-31T17%3A46%3A33Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/466805%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/omercs%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20src/ADAL.Common/WsTrustRequest.cs%3AL195-206%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-Pull 5021a1d4968aa64a1f235a2124fb94efd3b8bc53 src/ADAL.Common/WsTrustRequest.cs 163'></a> &nbsp;<b><a href='https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/223#discussion_r35997221'>File: src/ADAL.Common/WsTrustRequest.cs:L195-206</a></b>
- <a href='https://github.com/omercs'></a> typo for expiryTimeString
- [x] <a href='#crh-comment-Pull 5021a1d4968aa64a1f235a2124fb94efd3b8bc53 src/ADAL.Common/MexParser.cs 52'></a> &nbsp;<b><a href='https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/223#discussion_r35997813'>File: src/ADAL.Common/MexParser.cs:L85-110</a></b>
- <a href='https://github.com/aj-michael'></a> Why is this randomized?

<a href='https://www.codereviewhub.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/223?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/223?approve=1'></a>&nbsp;<a href='https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/223'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by brandwe** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/223'>Refresh</a>.<a href='#crh-update-none'></a>

File: src/ADAL.Common/WsTrustRequest.cs [hunk 1]
```diff
@@ -171,11 +195,11 @@ private static StringBuilder BuildSecurityHeader(UserCredential credential)
...
 
-                securityHeaderBuilder.AppendFormat(
-                    "<o:Security s:mustUnderstand='1' xmlns:o='http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd'><u:Timestamp u:Id='_0'><u:Created>{0}</u:Created><u:Expires>{1}</u:Expires></u:Timestamp>{2}</o:Security>", 
-                    currentTimeString, 
-                    expiryTimString, 
-                    messageCredentialsBuilder);
+                    securityHeaderBuilder.AppendFormat(
+                        "<o:Security s:mustUnderstand='1' xmlns:o='http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd'><u:Timestamp u:Id='_0'><u:Created>{0}</u:Created><u:Expires>{1}</u:Expires></u:Timestamp>{2}</o:Security>",
+                        currentTimeString,
+                        expiryTimString,
```
Comment (omercs) [inline src/ADAL.Common/WsTrustRequest.cs]: typo for expiryTimeString

File: tests/Test.ADAL.NET.Unit/NonInteractiveTests.cs [hunk 2]
```diff
@@ -115,7 +115,7 @@ public class NonInteractiveTests : AdalTestsBase
             UserRealmDiscoveryResponse userRealmResponse = await UserRealmDiscoveryResponse.CreateByDiscoveryAsync(context.Authenticator.UserRealmUri, federatedSts.ValidUserName, null);
             XDocument mexDocument = await FecthMexAsync(userRealmResponse.FederationMetadataUrl);
             Verify.IsNotNull(mexDocument);
-            Uri wsTrustAddress = MexParser.ExtractWsTrustAddressFromMex(mexDocument, UserAuthType.IntegratedAuth, null);
+            WsTrustAddress wsTrustAddress = MexParser.ExtractWsTrustAddressFromMex(mexDocument, UserAuthType.IntegratedAuth, null);
```
Comment (omercs) [inline tests/Test.ADAL.NET.Unit/NonInteractiveTests.cs]: Can you add test that extracts ws2005 endpoint when mex response does not have wstrust13 endpoint?

File: src/ADAL.Common/MexParser.cs [hunk 3]
```diff
@@ -70,20 +85,25 @@ internal class MexParser
...
+            MexPolicy policy = null;
             try
             {
                 Dictionary<string, MexPolicy> policies = ReadPolicies(mexDocument);
                 Dictionary<string, MexPolicy> bindings = ReadPolicyBindings(mexDocument, policies);
                 SetPolicyEndpointAddresses(mexDocument, bindings);
                 Random random = new Random();
-                MexPolicy policy = policies.Values.Where(p => p.Url != null && p.AuthType == userAuthType).OrderBy(p => random.Next()).FirstOrDefault();
+                //try ws-trust 1.3 first
+                policy = policies.Values.Where(p => p.Url != null && p.AuthType == userAuthType && p.Version == WsTrustVersion.WsTrust13).OrderBy(p => random.Next()).FirstOrDefault() ??
```
Comment (aj-michael) [inline src/ADAL.Common/MexParser.cs]: Why is this randomized?

Comment (kpanwar) [inline src/ADAL.Common/MexParser.cs]: because we do not make a prefernce of one over the other.

**Pull request merged**